actions/node_discover_and_add_interfaces_by_name_and_type.py

```diff
@@ -0,0 +1,111 @@
+# Licensed to the StackStorm, Inc ('StackStorm') under one or more
+# contributor license agreements.  See the NOTICE file distributed with
+# this work for additional information regarding copyright ownership.
+# The ASF licenses this file to You under the Apache License, Version 2.0
+# (the "License"); you may not use this file except in compliance with
+# the License.  You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from lib.actions import OrionBaseAction
+
+
+class NodeDiscoverAndAddInterfacesbyNameandType(OrionBaseAction):
+    def run(self, node, admin_up_required, interface_names, interface_type):
+        """
+        Discover and add interfaces on an Orion node
+
+        Args:
+           node: Node to discover and add interfaces on.
+           admin_up_required: Filter add interfaces by ifAdminStatus up or not
+           interface_names: List of ifName values to filter add interfaces
+           interface_type: List of ifType values to filter add interfaces
+
+        """
+        results = {'added': [], 'existing': [], 'removed': []}
+
+        self.connect()
+
+        orion_node = self.get_node(node)
+
+        if not orion_node.npm:
+            raise ValueError("Node not found")
+
+        Discoverdinterfaces = self.invoke('Orion.NPM.Interfaces',
+                                          'DiscoverInterfacesOnNode',
+                                          orion_node.npm_id)
+
+        add_interfaces = []
+        for interface in Discoverdinterfaces['DiscoveredInterfaces']:
```

Comment:
The wish for this Action was to allow us to add interfaces on a device based upon the name of the interface.  The problem in this approach was Orion generates a "Caption" for the interface when you invoke the "DiscoverInterfacesOnNode" verb and the "Caption" data that is generated is not consistent (it depends on if the SNMP includes ifAlias/ifDescr data or not) and it will often include an odd "middot" character so I is difficult to create and manage an input list filter based upon this data.  The ifName data is always consistent and simple to get added to a list, but it is not available to Orion until after the "DiscoverInterfacesOnNode" completes and the interface is added to the Node.  The loop logic here works around that limitation by adding all of the interfaces to the Node that match the types (ifType is returned in the data from the "DiscoverInterfacesOnNode" verb) and Admin Status  (ifAdminStatus is also returned from the "DiscoverInterfacesOnNode" verb) passed as input to the Action.  Once these interfaces have been added to the Node, the ifName data is then retrieved in a subsequent call and all of the interfaces that are returned are compared to the list passed as input to the script.  Any of the interfaces that were added to the Node that are not in the list are then deleted from the Node.

It is a convoluted process but it had to be that way to work around the way Orion executes the "DiscoverInterfacesOnNode" verb, the data that is returned from that execution, and the way it calculates the "Caption" which does not have a direct correlation to SNMP data values.